TypeScript typings

> I would love to have typescript typings for this, as it makes learning a new project (and using it correctly) much easier

From Agoric/SES#64

When I try to do a `import {lockdown} from 'ses'` I get: `Cannot find module 'ses'.` is that expected? Same file seems to work with `node -r esm` (same code that I would expect to work in typescript).

--- a/packages/compartment-mapper/README.md
+++ b/packages/compartment-mapper/README.md
@@ -363,8 +363,8 @@ The compartment map shape:
 type CompartmentMap = {
   tags: Tags,
   entry: Entry,
-  compartments: Object<CompartmentName, Compartment>,
-  realms: Object<RealmName, Realm>, // TODO
+  compartments: Record<CompartmentName, Compartment>,
+  realms: Record<RealmName, Realm>, // TODO
 };
 
 // Tags are the build tags for the compartment.
@@ -401,7 +401,7 @@ type Location = string;
 
 // ModuleMap describes modules available in the compartment
 // that do not correspond to source files in the same compartment.
-type ModuleMap = Object<InternalModuleSpecifier, Module>;
+type ModuleMap = Record<InternalModuleSpecifier, Module>;
 
 // Module describes a module in a compartment.
 type Module = CompartmentModule | FileModule | ExitModule;
@@ -464,7 +464,7 @@ type ExternalModuleSpecifier = string;
 // The compartment mapper adds `{"json": "json"}` for good measure in both
 // cases, although Node.js (as of version 0.14.5) does not support importing
 // JSON modules from ESM.
-type ParserMap = Object<Extension, Parser>;
+type ParserMap = Record<Extension, Parser>;
 
 // Extension is a file extension such as "js" for "main.js" or "" for "README".
 type Extension = string;
@@ -482,7 +482,7 @@ type Parser = "mjs" | "cjs" | "json";
 // a file that is an ECMAScript module, regardless of its extension.
 // This is the mechanism that allows the compartment mapper to respect that
 // behavior.
-type ModuleParserMap = Object<InternalModuleSpecifier, Parser>;
+type ModuleParserMap = Record<InternalModuleSpecifier, Parser>;
 
 // ScopeMap is a map from internal module specifier prefixes
 // like "dependency" or "@organization/dependency" to another
@@ -494,7 +494,7 @@ type ModuleParserMap = Object<InternalModuleSpecifier, Parser>;
 // to a link into some internal module of the foreign compartment.
 >> When the compartment mapper creates an archive, it captures all of the Modules
 >> explicitly and erases the scopes entry.
-type ScopeMap = Object<InternalModuleSpecifier, Scope>;
+type ScopeMap = Record<InternalModuleSpecifier, Scope>;
 
 // Scope describes the compartment to use for all ad-hoc
 // entries in the compartment's module map.

--- a/packages/compartment-mapper/src/archive.js
+++ b/packages/compartment-mapper/src/archive.js
@@ -1,3 +1,4 @@
+// @ts-check
 /* eslint no-shadow: 0 */
 
 import { writeZip } from './zip.js';
@@ -10,11 +11,21 @@ import * as json from './json.js';
 
 const encoder = new TextEncoder();
 
+/**
+ * @param {string} rel - a relative URL
+ * @param {string} abs - a fully qualified URL
+ * @returns {string}
+ */
 const resolveLocation = (rel, abs) => new URL(rel, abs).toString();
 
 const { entries, fromEntries, values } = Object;
 
+/**
+ * @param {Record<string, CompartmentDescriptor>} compartments
+ * @returns {Record<string, string>} map from old to new compartment names.
+ */
 const renameCompartments = compartments => {
+  /** @type {Record<string, string>} */
   const renames = {};
   let n = 0;
   for (const [name, compartment] of entries(compartments)) {
@@ -25,12 +36,18 @@ const renameCompartments = compartments => {
   return renames;
 };
 
+/**
+ * @param {Record<string, CompartmentDescriptor>} compartments
+ * @param {Sources} sources
+ * @param {Record<string, string>} renames
+ */
 const translateCompartmentMap = (compartments, sources, renames) => {
   const result = {};
   for (const [name, compartment] of entries(compartments)) {
     const { label } = compartment;
 
     // rename module compartments
+    /** @type {Record<string, ModuleDescriptor>} */
     const modules = {};
     for (const [name, module] of entries(compartment.modules || {})) {
       const compartment = module.compartment
@@ -65,26 +82,42 @@ const translateCompartmentMap = (compartments, sources, renames) => {
   return result;
 };
 
+/**
+ * @param {Sources} sources
+ * @param {Record<string, string>} renames
+ * @returns {Sources}
+ */
 const renameSources = (sources, renames) => {
   return fromEntries(
     entries(sources).map(([name, compartment]) => [renames[name], compartment]),
   );
 };
 
+/**
+ * @param {ArchiveWriter} archive
+ * @param {Sources} sources
+ */
 const addSourcesToArchive = async (archive, sources) => {
   for (const [compartment, modules] of entries(sources)) {
     const compartmentLocation = resolveLocation(`${compartment}/`, 'file:///');
     for (const { location, bytes } of values(modules)) {
-      const moduleLocation = resolveLocation(location, compartmentLocation);
-      const path = new URL(moduleLocation).pathname.slice(1); // elide initial "/"
-      if (bytes !== undefined) {
-        // eslint-disable-next-line no-await-in-loop
-        await archive.write(path, bytes);
+      if (location !== undefined) {
+        const moduleLocation = resolveLocation(location, compartmentLocation);
+        const path = new URL(moduleLocation).pathname.slice(1); // elide initial "/"
+        if (bytes !== undefined) {
+          // eslint-disable-next-line no-await-in-loop
+          await archive.write(path, bytes);
+        }
       }
     }
   }
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} moduleLocation
+ * @returns {Promise<Uint8Array>}
+ */
 export const makeArchive = async (read, moduleLocation) => {
   const {
     packageLocation,
@@ -93,6 +126,9 @@ export const makeArchive = async (read, moduleLocation) => {
     moduleSpecifier,
   } = await search(read, moduleLocation);
 
+  /** @type {Set<string>} */
+  const tags = new Set();
+
   const packageDescriptor = json.parse(
     packageDescriptorText,
     packageDescriptorLocation,
@@ -100,7 +136,7 @@ export const makeArchive = async (read, moduleLocation) => {
   const compartmentMap = await compartmentMapForNodeModules(
     read,
     packageLocation,
-    [],
+    tags,
     packageDescriptor,
     moduleSpecifier,
   );
@@ -109,6 +145,7 @@ export const makeArchive = async (read, moduleLocation) => {
     compartments,
     entry: { compartment: entryCompartmentName, module: entryModuleSpecifier },
   } = compartmentMap;
+  /** @type {Sources} */
   const sources = {};
 
   const makeImportHook = makeImportHookMaker(
@@ -155,6 +192,12 @@ export const makeArchive = async (read, moduleLocation) => {
   return archive.snapshot();
 };
 
+/**
+ * @param {WriteFn} write
+ * @param {ReadFn} read
+ * @param {string} archiveLocation
+ * @param {string} moduleLocation
+ */
 export const writeArchive = async (
   write,
   read,

--- a/packages/compartment-mapper/src/assemble.js
+++ b/packages/compartment-mapper/src/assemble.js
@@ -1,3 +1,6 @@
+// @ts-check
+/// <reference types="ses" />
+
 import { resolve } from './node-module-specifier.js';
 import { mapParsers } from './parse.js';
 
@@ -8,10 +11,16 @@ const defaultCompartment = Compartment;
 // q, as in quote, for strings in error messages.
 const q = JSON.stringify;
 
-// For a full, absolute module specifier like "dependency",
-// produce the module specifier in the dependency, like ".".
-// For a deeper path like "@org/dep/aux" and a prefix like "@org/dep", produce
-// "./aux".
+/**
+ * For a full, absolute module specifier like "dependency",
+ * produce the module specifier in the dependency, like ".".
+ * For a deeper path like "@org/dep/aux" and a prefix like "@org/dep", produce
+ * "./aux".
+ *
+ * @param {string} moduleSpecifier
+ * @param {string} prefix
+ * @returns {string=}
+ */
 const trimModuleSpecifierPrefix = (moduleSpecifier, prefix) => {
   if (moduleSpecifier === prefix) {
     return '.';
@@ -22,14 +31,23 @@ const trimModuleSpecifierPrefix = (moduleSpecifier, prefix) => {
   return undefined;
 };
 
-// `makeModuleMapHook` generates a `moduleMapHook` for the `Compartment`
-// constructor, suitable for Node.js style packages where any module in the
-// package might be imported.
-// Since searching for all of these modules up front is either needlessly
-// costly (on a file system) or impossible (from a web service), we
-// let the import graph guide our search.
-// Any module specifier with an absolute prefix should be captured by
-// the `moduleMap` or `moduleMapHook`.
+/**
+ * `makeModuleMapHook` generates a `moduleMapHook` for the `Compartment`
+ * constructor, suitable for Node.js style packages where any module in the
+ * package might be imported.
+ * Since searching for all of these modules up front is either needlessly
+ * costly (on a file system) or impossible (from a web service), we
+ * let the import graph guide our search.
+ * Any module specifier with an absolute prefix should be captured by
+ * the `moduleMap` or `moduleMapHook`.
+ *
+ * @param {Record<string, Compartment>} compartments
+ * @param {string} compartmentName
+ * @param {Record<string, ModuleDescriptor>} moduleDescriptors
+ * @param {Record<string, ModuleDescriptor>} scopeDescriptors
+ * @param {Record<string, string>} exitModules
+ * @returns {ModuleMapHook | undefined}
+ */
 const makeModuleMapHook = (
   compartments,
   compartmentName,
@@ -37,6 +55,10 @@ const makeModuleMapHook = (
   scopeDescriptors,
   exitModules,
 ) => {
+  /**
+   * @param {string} moduleSpecifier
+   * @returns {string | Object | undefined}
+   */
   const moduleMapHook = moduleSpecifier => {
     const moduleDescriptor = moduleDescriptors[moduleSpecifier];
     if (moduleDescriptor !== undefined) {
@@ -82,6 +104,11 @@ const makeModuleMapHook = (
 
       if (foreignModuleSpecifier !== undefined) {
         const { compartment: foreignCompartmentName } = scopeDescriptor;
+        if (foreignCompartmentName === undefined) {
+          throw new Error(
+            `Cannot import from scope ${scopePrefix} due to missing "compartment" property`,
+          );
+        }
         const foreignCompartment = compartments[foreignCompartmentName];
         if (foreignCompartment === undefined) {
           throw new Error(
@@ -115,16 +142,21 @@ const makeModuleMapHook = (
   return moduleMapHook;
 };
 
-// Assemble a DAG of compartments as declared in a compartment map starting at
-// the named compartment and building all compartments that it depends upon,
-// recursively threading the modules exported by one compartment into the
-// compartment that imports them.
-// Returns the root of the compartment DAG.
-// Does not load or execute any modules.
-// Uses makeImportHook with the given "location" string of each compartment in
-// the DAG.
-// Passes the given globals and external modules into the root compartment
-// only.
+/**
+ * Assemble a DAG of compartments as declared in a compartment map starting at
+ * the named compartment and building all compartments that it depends upon,
+ * recursively threading the modules exported by one compartment into the
+ * compartment that imports them.
+ * Returns the root of the compartment DAG.
+ * Does not load or execute any modules.
+ * Uses makeImportHook with the given "location" string of each compartment in
+ * the DAG.
+ * Passes the given globals and external modules into the root compartment
+ * only.
+ *
+ * @param {CompartmentMapDescriptor} compartmentMap
+ * @param {AssemblyOptions} options
+ */
 export const assemble = (
   { entry, compartments: compartmentDescriptors },
   {
@@ -139,6 +171,7 @@ export const assemble = (
 ) => {
   const { compartment: entryCompartmentName } = entry;
 
+  /** @type {Record<string, Compartment>} */
   const compartments = {};
   for (const [compartmentName, compartmentDescriptor] of entries(
     compartmentDescriptors,

--- a/packages/compartment-mapper/src/extension.js
+++ b/packages/compartment-mapper/src/extension.js
@@ -1,6 +1,13 @@
-// `parseExtension` returns the file extension for the given URL, or an empty
-// string if the path has no extension.
-// Exported for tests.
+// @ts-check
+
+/**
+ * `parseExtension` returns the file extension for the given URL, or an empty
+ * string if the path has no extension.
+ * Exported for tests.
+ *
+ * @param {string} location
+ * @returns {string}
+ */
 export const parseExtension = location => {
   const lastSlash = location.lastIndexOf('/');
   if (lastSlash < 0) {

--- a/packages/compartment-mapper/src/import-archive.js
+++ b/packages/compartment-mapper/src/import-archive.js
@@ -1,3 +1,4 @@
+// @ts-check
 /* eslint no-shadow: "off" */
 
 import { readZip } from './zip.js';
@@ -10,14 +11,35 @@ const q = JSON.stringify;
 
 const decoder = new TextDecoder();
 
+/**
+ * @callback ArchiveImportHookMaker
+ * @param {string} packageLocation
+ * @returns {ImportHook}
+ */
+
+/**
+ * @param {ArchiveReader} archive
+ * @param {Record<string, CompartmentDescriptor>} compartments
+ * @param {string} archiveLocation
+ * @returns {ArchiveImportHookMaker}
+ */
 const makeArchiveImportHookMaker = (archive, compartments, archiveLocation) => {
   // per-assembly:
+  /** @type {ArchiveImportHookMaker} */
   const makeImportHook = packageLocation => {
     // per-compartment:
     const { modules } = compartments[packageLocation];
+    /** @type {ImportHook} */
     const importHook = async moduleSpecifier => {
       // per-module:
       const module = modules[moduleSpecifier];
+      if (module.parser === undefined) {
+        throw new Error(
+          `Cannot parse module ${q(moduleSpecifier)} in package ${q(
+            packageLocation,
+          )} in archive ${q(archiveLocation)}`,
+        );
+      }
       const parse = parserForLanguage[module.parser];
       if (parse === undefined) {
         throw new Error(
@@ -41,13 +63,28 @@ const makeArchiveImportHookMaker = (archive, compartments, archiveLocation) => {
   return makeImportHook;
 };
 
+/**
+ * @param {Uint8Array} archiveBytes
+ * @param {string} archiveLocation
+ * @returns {Promise<Application>}
+ */
 export const parseArchive = async (archiveBytes, archiveLocation) => {
   const archive = await readZip(archiveBytes, archiveLocation);
 
   const compartmentMapBytes = await archive.read('compartment-map.json');
   const compartmentMapText = decoder.decode(compartmentMapBytes);
-  const compartmentMap = json.parse(compartmentMapText, 'compartment-map.json');
+  const compartmentMap = /** @type {CompartmentMapDescriptor} */ (json.parse(
+    compartmentMapText,
+    'compartment-map.json',
+  ));
 
+  // TODO validate compartmentMap instead of leaning hard on the above type
+  // assertion.
+
+  /**
+   * @param {ExecuteOptions} options
+   * @returns {Promise<Object>}
+   */
   const execute = options => {
     const {
       globals,
@@ -83,11 +120,22 @@ export const parseArchive = async (archiveBytes, archiveLocation) => {
   return { import: execute };
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} archiveLocation
+ * @returns {Promise<Application>}
+ */
 export const loadArchive = async (read, archiveLocation) => {
   const archiveBytes = await read(archiveLocation);
   return parseArchive(archiveBytes, archiveLocation);
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} archiveLocation
+ * @param {ExecuteOptions} options
+ * @returns {Promise<Object>}
+ */
 export const importArchive = async (read, archiveLocation, options) => {
   const archive = await loadArchive(read, archiveLocation);
   // Call import by property to bypass SES censoring for dynamic import.

--- a/packages/compartment-mapper/src/import-hook.js
+++ b/packages/compartment-mapper/src/import-hook.js
@@ -1,3 +1,4 @@
+// @ts-check
 import { parseExtension } from './extension.js';
 
 // q, as in quote, for quoting strings in error messages.
@@ -7,8 +8,20 @@ const decoder = new TextDecoder();
 
 const { freeze } = Object;
 
+/**
+ * @param {string} rel - a relative URL
+ * @param {string} abs - a fully qualified URL
+ * @returns {string}
+ */
 const resolveLocation = (rel, abs) => new URL(rel, abs).toString();
 
+/**
+ * @param {ReadFn} read
+ * @param {string} baseLocation
+ * @param {Sources} sources
+ * @param {Record<string, CompartmentDescriptor>} compartments
+ * @returns {ImportHookMaker}
+ */
 export const makeImportHookMaker = (
   read,
   baseLocation,
@@ -16,6 +29,7 @@ export const makeImportHookMaker = (
   compartments = {},
 ) => {
   // per-assembly:
+  /** @type {ImportHookMaker} */
   const makeImportHook = (packageLocation, parse) => {
     // per-compartment:
     packageLocation = resolveLocation(packageLocation, baseLocation);
@@ -23,6 +37,7 @@ export const makeImportHookMaker = (
     sources[packageLocation] = packageSources;
     const { modules = {} } = compartments[packageLocation] || {};
 
+    /** @type {ImportHook} */
     const importHook = async moduleSpecifier => {
       // per-module:
 
@@ -76,16 +91,23 @@ export const makeImportHookMaker = (
             packageLocation,
           );
           const { parser } = envelope;
-          let { record } = envelope;
+          const { record: concreteRecord } = envelope;
 
           // Facilitate a redirect if the returned record has a different
           // module specifier than the requested one.
+          /** @type {StaticModuleType} */
+          let record;
           if (candidateSpecifier !== moduleSpecifier) {
             modules[moduleSpecifier] = {
               module: candidateSpecifier,
               compartment: packageLocation,
             };
-            record = { record, specifier: candidateSpecifier };
+            record = {
+              record: concreteRecord,
+              specifier: candidateSpecifier,
+            };
+          } else {
+            record = concreteRecord;
           }
 
           const packageRelativeLocation = moduleLocation.slice(
@@ -105,7 +127,7 @@ export const makeImportHookMaker = (
         `Cannot find file for internal module ${q(
           moduleSpecifier,
         )} (with candidates ${candidates
-          .map(q)
+          .map(x => q(x))
           .join(', ')}) in package ${packageLocation}`,
       );
     };

--- a/packages/compartment-mapper/src/import.js
+++ b/packages/compartment-mapper/src/import.js
@@ -1,3 +1,4 @@
+// @ts-check
 /* eslint no-shadow: "off" */
 
 import { compartmentMapForNodeModules } from './node-modules.js';
@@ -6,6 +7,11 @@ import { assemble } from './assemble.js';
 import { makeImportHookMaker } from './import-hook.js';
 import * as json from './json.js';
 
+/**
+ * @param {ReadFn} read
+ * @param {string} moduleLocation
+ * @returns {Promise<Application>}
+ */
 export const loadLocation = async (read, moduleLocation) => {
   const {
     packageLocation,
@@ -18,11 +24,14 @@ export const loadLocation = async (read, moduleLocation) => {
     packageDescriptorText,
     packageDescriptorLocation,
   );
+  /** @type {Set<string>} */
+  const tags = new Set();
   const compartmentMap = await compartmentMapForNodeModules(
     read,
     packageLocation,
-    [],
+    tags,
     packageDescriptor,
+    moduleSpecifier,
   );
 
   const execute = async (options = {}) => {
@@ -52,6 +61,13 @@ export const loadLocation = async (read, moduleLocation) => {
   return { import: execute };
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} moduleLocation
+ * @param {ExecuteOptions} options
+ * @returns {Promise<Object>} the object of the imported modules exported
+ * names.
+ */
 export const importLocation = async (read, moduleLocation, options = {}) => {
   const application = await loadLocation(read, moduleLocation);
   // Call import by property to bypass SES censoring for dynamic import.

--- a/packages/compartment-mapper/src/infer-exports.js
+++ b/packages/compartment-mapper/src/infer-exports.js
@@ -1,7 +1,14 @@
+// @ts-check
+
 import { join, relativize } from './node-module-specifier.js';
 
 const { entries, fromEntries } = Object;
 
+/**
+ * @param {string} name - the name of the referrer package.
+ * @param {Object} exports - the `exports` field from a package.json
+ * @yields {[string, string]}
+ */
 function* interpretBrowserExports(name, exports) {
   if (typeof exports === 'string') {
     yield [name, relativize(exports)];
@@ -17,6 +24,13 @@ function* interpretBrowserExports(name, exports) {
   }
 }
 
+/**
+ * @param {string} name - the name of the referrer package.
+ * @param {Object} exports - the `exports` field from a package.json.
+ * @param {Set<string>} tags - build tags about the target environment
+ * for selecting relevant exports, e.g., "browser" or "node".
+ * @yields {[string, string]}
+ */
 function* interpretExports(name, exports, tags) {
   if (typeof exports === 'string') {
     yield [name, relativize(exports)];
@@ -36,14 +50,28 @@ function* interpretExports(name, exports, tags) {
   }
 }
 
-// Given an unpacked `package.json`, generate a series of `[name, target]`
-// pairs to represent what this package exports. `name` is what the
-// caller/importer asked for (for example, the `ses` in `import { stuff } from
-// 'ses'`, or the `ses/deeper` in `import { stuff } from 'ses/deeper'`).
-// `target` is the path relative to the imported package's root: frequently
-// `./index.js` or `./src/index.js` or (for a deep import) `./src/deeper.js`.
-// There may be multiple pairs for a single `name`, but they will be yielded in
-// ascending priority order, and the caller should use the last one that exists.
+/**
+ * Given an unpacked `package.json`, generate a series of `[name, target]`
+ * pairs to represent what this package exports. `name` is what the
+ * caller/importer asked for (for example, the `ses` in `import { stuff } from
+ * 'ses'`, or the `ses/deeper` in `import { stuff } from 'ses/deeper'`).
+ * `target` is the path relative to the imported package's root: frequently
+ * `./index.js` or `./src/index.js` or (for a deep import) `./src/deeper.js`.
+ * There may be multiple pairs for a single `name`, but they will be yielded in
+ * ascending priority order, and the caller should use the last one that exists.
+ *
+ * @param {Object} packageDescriptor - the parsed body of a package.json file.
+ * @param {string} packageDescriptor.name
+ * @param {string} packageDescriptor.main
+ * @param {string} [packageDescriptor.module]
+ * @param {string} [packageDescriptor.browser]
+ * @param {Object} [packageDescriptor.exports]
+ * @param {Set<string>} tags - build tags about the target environment
+ * for selecting relevant exports, e.g., "browser" or "node".
+ * @param {Record<string, ParserDescriptor>} types - an object to populate
+ * with any recognized module's type, if implied by a tag.
+ * @yields {[string, string]}
+ */
 export function* inferExportsEntries(
   { name, main, module, browser, exports },
   tags,
@@ -73,20 +101,29 @@ export function* inferExportsEntries(
   // modules, taking care to exclude node_modules.
 }
 
-// inferExports reads a package.json (package descriptor) and constructs a map
-// of all the modules that package exports.
-// The keys are the module specifiers for the module map of any package that
-// depends upon this one, like `semver` for the main module of the `semver`
-// package.
-// The values are the corresponding module specifiers in the dependency
-// package's module map, like `./index.js`.
-//
-// TODO When a package does not supply the `exports` property, this function
-// needs to infer that all JavaScript modules in the package are exported.
-// Most packages will need this.
-// This function can remain synchronous if we pre-populate a file manifest for
-// every package.
-// That manifest will also prove useful for resolving aliases, like the
-// implicit index.js modules within a package.
+/**
+ * inferExports reads a package.json (package descriptor) and constructs a map
+ * of all the modules that package exports.
+ * The keys are the module specifiers for the module map of any package that
+ * depends upon this one, like `semver` for the main module of the `semver`
+ * package.
+ * The values are the corresponding module specifiers in the dependency
+ * package's module map, like `./index.js`.
+ *
+ * TODO When a package does not supply the `exports` property, this function
+ * needs to infer that all JavaScript modules in the package are exported.
+ * Most packages will need this.
+ * This function can remain synchronous if we pre-populate a file manifest for
+ * every package.
+ * That manifest will also prove useful for resolving aliases, like the
+ * implicit index.js modules within a package.
+ *
+ * @param {Object} descriptor - the parsed body of a package.json file.
+ * @param {Set<string>} tags - build tags about the target environment
+ * for selecting relevant exports, e.g., "browser" or "node".
+ * @param {Record<string, ParserDescriptor>} types - an object to populate
+ * with any recognized module's type, if implied by a tag.
+ * @returns {Record<string, string>}
+ */
 export const inferExports = (descriptor, tags, types) =>
   fromEntries(inferExportsEntries(descriptor, tags, types));

--- a/packages/compartment-mapper/src/json.js
+++ b/packages/compartment-mapper/src/json.js
@@ -1,3 +1,12 @@
+// @ts-check
+
+/**
+ * Parses JSON and, if necessary, throws exceptions that include the location
+ * of the offending file.
+ *
+ * @param {string} source
+ * @param {string} location
+ */
 export const parse = (source, location) => {
   try {
     return JSON.parse(source);

--- a/packages/compartment-mapper/src/main.js
+++ b/packages/compartment-mapper/src/main.js
@@ -1,3 +1,4 @@
+// @ts-check
 export { loadLocation, importLocation } from './import.js';
 export { makeArchive, writeArchive } from './archive.js';
 export { parseArchive, loadArchive, importArchive } from './import-archive.js';

--- a/packages/compartment-mapper/src/node-module-specifier.js
+++ b/packages/compartment-mapper/src/node-module-specifier.js
@@ -1,13 +1,24 @@
+// @ts-check
+
 // q, as in quote, for error messages.
 const q = JSON.stringify;
 
-// Advances a partial module specifier solution by following the path
-// components in the given problem.
-// The problem may not produce a path that escapes the solution, that is, the
-// problem may not traverse up from an empty solution.
-// `Solve` returns false if the problem attempts to escape.
-// Advanding a partial solution is the core of `resolve`, `join`, and
-// `relativize`, which have different invariants.
+/**
+ * Advances a partial module specifier solution by following the path
+ * components in the given problem.
+ * The problem may not produce a path that escapes the solution, that is, the
+ * problem may not traverse up from an empty solution.
+ * `Solve` returns false if the problem attempts to escape.
+ * Advanding a partial solution is the core of `resolve`, `join`, and
+ * `relativize`, which have different invariants.
+ *
+ * @param {Array<string>} solution - fully resolved path components, including
+ * any from a prior path resolution initially.
+ * @param {Array<string>} problem - partially resolved path components, that
+ * is, including '.' and '..' components.
+ * @returns {boolean} whether the solver terminated early because of a
+ * nonsensical attempt to traverse above the root directory.
+ */
 const solve = (solution, problem) => {
   for (const part of problem) {
     if (part === '.' || part === '') {
@@ -24,15 +35,21 @@ const solve = (solution, problem) => {
   return true;
 };
 
-// `Resolve` computes the full module specifier for a given imported module specifier
-// relative to the referrer module specifier.
-// In Node.js compartments, the referrer must be an internal module specifier
-// in the context of a compartment, and all internal module specifiers begin
-// with a "." path component.
-// The referent may be either internal or external.
-// In Node.js, fully qualified paths are valid module specifiers, but these
-// paths that begin with / are disallowed as they could be used to defeat
-// compartment containment.
+/**
+ * `Resolve` computes the full module specifier for a given imported module specifier
+ * relative to the referrer module specifier.
+ * In Node.js compartments, the referrer must be an internal module specifier
+ * in the context of a compartment, and all internal module specifiers begin
+ * with a "." path component.
+ * The referent may be either internal or external.
+ * In Node.js, fully resolved paths are valid module specifiers, but these
+ * paths that begin with / are disallowed as they could be used to defeat
+ * compartment containment.
+ *
+ * @param {string} spec - a path to resolve.
+ * @param {string} referrer - the fully resolved path of referrer module.
+ * @returns {string} the fully resolved path.
+ */
 export const resolve = (spec, referrer) => {
   spec = String(spec || '');
   referrer = String(referrer || '');
@@ -66,13 +83,19 @@ export const resolve = (spec, referrer) => {
   return solution.join('/');
 };
 
-// To construct a module map from a node_modules package, inter-package linkage
-// requires connecting a full base module specifier like "dependency-package"
-// to the other package's full internal module specifier like "." or
-// "./utility", to form a local full module specifier like "dependency-package"
-// or "dependency-package/utility".
-// This type of join may assert that the base is absolute and the referrent is
-// relative.
+/**
+ * To construct a module map from a node_modules package, inter-package linkage
+ * requires connecting a full base module specifier like "dependency-package"
+ * to the other package's full internal module specifier like "." or
+ * "./utility", to form a local full module specifier like "dependency-package"
+ * or "dependency-package/utility".
+ * This type of join may assert that the base is absolute and the referrent is
+ * relative.
+ *
+ * @param {string} base - the fully resolved path of a module.
+ * @param {string} spec - the partially resolved path of another module.
+ * @returns {string} the fully resolved path of the specified module.
+ */
 export const join = (base, spec) => {
   spec = String(spec || '');
   base = String(base || '');
@@ -102,10 +125,17 @@ export const join = (base, spec) => {
   return [base, ...solution].join('/');
 };
 
-// Relativize turns absolute identifiers into relative identifiers.
-// In package.json, internal module identifiers can be either relative or
-// absolute, but compartments backed by node_modules always use relative module
-// specifiers for internal linkage.
+/**
+ * Relativize turns absolute identifiers into relative identifiers.
+ * In package.json, internal module identifiers can be either relative or
+ * absolute, but compartments backed by node_modules always use relative module
+ * specifiers for internal linkage.
+ *
+ * @param {string} spec - a module specifier that of a local module, that might
+ * be erroneously framed without an initial '.' path component.
+ * @returns {string} the idempotent module specifier, ensured to begin with
+ * '.'.
+ */
 export const relativize = spec => {
   spec = String(spec || '');
 

--- a/packages/compartment-mapper/src/node-modules.js
+++ b/packages/compartment-mapper/src/node-modules.js
@@ -1,4 +1,27 @@
+// @ts-check
 /* eslint no-shadow: 0 */
+
+/**
+ * The graph is an intermediate object model that the functions of this module
+ * build by exploring the `node_modules` tree dropped by tools like npm and
+ * consumed by tools like Node.js.
+ * This gets translated finally into a compartment map.
+ *
+ * @typedef {Record<string, Node>} Graph
+ */
+
+/**
+ * @typedef {Object} Node
+ * @property {string} label
+ * @property {boolean} explicit
+ * @property {Record<string, string>} exports
+ * @property {Record<string, string>} dependencies - from module name to
+ * location in storage.
+ * @property {Record<string, ParserDescriptor>} parsers - the parser for
+ * modules based on their extension.
+ * @property {Record<string, ParserDescriptor>} types - the parser for specific
+ * modules.
+ */
 
 import { inferExports } from './infer-exports.js';
 import * as json from './json.js';
@@ -10,8 +33,17 @@ const decoder = new TextDecoder();
 // q, as in quote, for enquoting strings in error messages.
 const q = JSON.stringify;
 
+/**
+ * @param {string} rel - a relative URL
+ * @param {string} abs - a fully qualified URL
+ * @returns {string}
+ */
 const resolveLocation = (rel, abs) => new URL(rel, abs).toString();
 
+/**
+ * @param {string} location
+ * @returns {string}
+ */
 const basename = location => {
   const { pathname } = new URL(location);
   const index = pathname.lastIndexOf('/');
@@ -21,6 +53,11 @@ const basename = location => {
   return pathname.slice(index + 1);
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} packageLocation
+ * @returns {Promise<Object>}
+ */
 const readDescriptor = async (read, packageLocation) => {
   const descriptorLocation = resolveLocation('package.json', packageLocation);
   const descriptorBytes = await read(descriptorLocation).catch(
@@ -34,6 +71,12 @@ const readDescriptor = async (read, packageLocation) => {
   return descriptor;
 };
 
+/**
+ * @param {Record<string, Object>} memo
+ * @param {ReadFn} read
+ * @param {string} packageLocation
+ * @returns {Promise<Object>}
+ */
 const readDescriptorWithMemo = async (memo, read, packageLocation) => {
   let promise = memo[packageLocation];
   if (promise !== undefined) {
@@ -44,12 +87,28 @@ const readDescriptorWithMemo = async (memo, read, packageLocation) => {
   return promise;
 };
 
-// findPackage behaves as Node.js to find third-party modules by searching
-// parent to ancestor directories for a `node_modules` directory that contains
-// the name.
-// Node.js does not actually require these to be packages, but in practice,
-// these are the locations that pakcage managers drop a package so Node.js can
-// find it efficiently.
+/**
+ * @callback ReadDescriptorFn
+ * @param {string} packageLocation
+ * @returns {Promise<Object>}
+ */
+
+/**
+ * findPackage behaves as Node.js to find third-party modules by searching
+ * parent to ancestor directories for a `node_modules` directory that contains
+ * the name.
+ * Node.js does not actually require these to be packages, but in practice,
+ * these are the locations that pakcage managers drop a package so Node.js can
+ * find it efficiently.
+ *
+ * @param {ReadDescriptorFn} readDescriptor
+ * @param {string} directory
+ * @param {string} name
+ * @returns {Promise<{
+ *   packageLocation: string,
+ *   packageDescriptor: Object,
+ * } | undefined>}
+ */
 const findPackage = async (readDescriptor, directory, name) => {
   for (;;) {
     const packageLocation = resolveLocation(`node_modules/${name}/`, directory);
@@ -81,6 +140,11 @@ const uncontroversialParsers = { mjs: 'mjs', json: 'json' };
 const commonParsers = uncontroversialParsers;
 const moduleParsers = { js: 'mjs', ...uncontroversialParsers };
 
+/**
+ * @param {Object} descriptor
+ * @param {string} location
+ * @returns {Record<string, string>}
+ */
 const inferParsers = (descriptor, location) => {
   const { type, parsers } = descriptor;
   if (parsers !== undefined) {
@@ -117,15 +181,25 @@ const inferParsers = (descriptor, location) => {
   return commonParsers;
 };
 
-// graphPackage and gatherDependency are mutually recursive functions that
-// gather the metadata for a package and its transitive dependencies.
-// The keys of the graph are the locations of the package descriptors.
-// The metadata include a label (which is informative and not necessarily
-// unique), the location of each shallow dependency, and names of the modules
-// that the package exports.
-
+/**
+ * graphPackage and gatherDependency are mutually recursive functions that
+ * gather the metadata for a package and its transitive dependencies.
+ * The keys of the graph are the locations of the package descriptors.
+ * The metadata include a label (which is informative and not necessarily
+ * unique), the location of each shallow dependency, and names of the modules
+ * that the package exports.
+ *
+ * @param {string} name
+ * @param {ReadDescriptorFn} readDescriptor
+ * @param {Graph} graph
+ * @param {Object} packageDetails
+ * @param {string} packageDetails.packageLocation
+ * @param {Object} packageDetails.packageDescriptor
+ * @param {Set<string>} tags
+ * @returns {Promise<undefined>}
+ */
 const graphPackage = async (
-  name = '',
+  name,
   readDescriptor,
   graph,
   { packageLocation, packageDescriptor },
@@ -143,8 +217,9 @@ const graphPackage = async (
   }
 
   const result = {};
-  graph[packageLocation] = result;
+  graph[packageLocation] = /** @type {Node} */ (result);
 
+  /** @type {Record<string, string>} */
   const dependencies = {};
   const children = [];
   for (const name of keys(packageDescriptor.dependencies || {})) {
@@ -163,6 +238,7 @@ const graphPackage = async (
   }
 
   const { version = '', exports } = packageDescriptor;
+  /** @type {Record<string, ParserDescriptor>} */
   const types = {};
 
   Object.assign(result, {
@@ -174,9 +250,18 @@ const graphPackage = async (
     parsers: inferParsers(packageDescriptor, packageLocation),
   });
 
-  return Promise.all(children);
+  await Promise.all(children);
+  return undefined;
 };
 
+/**
+ * @param {ReadDescriptorFn} readDescriptor
+ * @param {Graph} graph - the partially build graph.
+ * @param {Record<string, string>} dependencies
+ * @param {string} packageLocation - location of the package of interest.
+ * @param {string} name - name of the package of interest.
+ * @param {Set<string>} tags
+ */
 const gatherDependency = async (
   readDescriptor,
   graph,
@@ -193,13 +278,21 @@ const gatherDependency = async (
   await graphPackage(name, readDescriptor, graph, dependency, tags);
 };
 
-// graphPackages returns a graph whose keys are nominally URLs, one per
-// package, with values that are label: (an informative Compartment name, built
-// as ${name}@${version}), dependencies: (a list of URLs), and exports: (an
-// object whose keys are the thing being imported, and the values are the names
-// of the matching module, relative to the containing package's root, that is,
-// the URL that was used as the key of graph).
-// The URLs in dependencies will all exist as other keys of graph.
+/**
+ * graphPackages returns a graph whose keys are nominally URLs, one per
+ * package, with values that are label: (an informative Compartment name, built
+ * as ${name}@${version}), dependencies: (a list of URLs), and exports: (an
+ * object whose keys are the thing being imported, and the values are the names
+ * of the matching module, relative to the containing package's root, that is,
+ * the URL that was used as the key of graph).
+ * The URLs in dependencies will all exist as other keys of graph.
+ *
+ * @param {ReadFn} read
+ * @param {string} packageLocation - location of the main package.
+ * @param {Set<string>} tags
+ * @param {Object} mainPackageDescriptor - the parsed contents of the main
+ * package.json, which was already read when searching for the package.json.
+ */
 const graphPackages = async (
   read,
   packageLocation,
@@ -207,6 +300,10 @@ const graphPackages = async (
   mainPackageDescriptor,
 ) => {
   const memo = create(null);
+  /**
+   * @param {string} packageLocation
+   * @returns {Promise<Object>}
+   */
   const readDescriptor = packageLocation =>
     readDescriptorWithMemo(memo, read, packageLocation);
 
@@ -238,9 +335,24 @@ const graphPackages = async (
   return graph;
 };
 
-// translateGraph converts the graph returned by graph packages (above) into a
-// compartment map.
-const translateGraph = (entryPackageLocation, entryModuleSpecifier, graph) => {
+/**
+ * translateGraph converts the graph returned by graph packages (above) into a
+ * compartment map.
+ *
+ * @param {string} entryPackageLocation
+ * @param {string} entryModuleSpecifier
+ * @param {Graph} graph
+ * @param {Set<string>} tags - build tags about the target environment
+ * for selecting relevant exports, e.g., "browser" or "node".
+ * @returns {CompartmentMapDescriptor}
+ */
+const translateGraph = (
+  entryPackageLocation,
+  entryModuleSpecifier,
+  graph,
+  tags,
+) => {
+  /** @type {Record<string, CompartmentDescriptor>} */
   const compartments = {};
 
   // For each package, build a map of all the external modules the package can
@@ -256,7 +368,9 @@ const translateGraph = (entryPackageLocation, entryModuleSpecifier, graph) => {
     packageLocation,
     { label, dependencies, parsers, types },
   ] of entries(graph)) {
+    /** @type {Record<string, ModuleDescriptor>} */
     const modules = {};
+    /** @type {Record<string, ScopeDescriptor>} */
     const scopes = {};
     for (const [dependencyName, packageLocation] of entries(dependencies)) {
       const { exports, explicit } = graph[packageLocation];
@@ -283,6 +397,7 @@ const translateGraph = (entryPackageLocation, entryModuleSpecifier, graph) => {
   }
 
   return {
+    tags: [...tags],
     entry: {
       compartment: entryPackageLocation,
       module: entryModuleSpecifier,
@@ -291,6 +406,14 @@ const translateGraph = (entryPackageLocation, entryModuleSpecifier, graph) => {
   };
 };
 
+/**
+ * @param {ReadFn} read
+ * @param {string} packageLocation
+ * @param {Set<string>} tags
+ * @param {Object} packageDescriptor
+ * @param {string} moduleSpecifier
+ * @returns {Promise<CompartmentMapDescriptor>}
+ */
 export const compartmentMapForNodeModules = async (
   read,
   packageLocation,
@@ -304,5 +427,5 @@ export const compartmentMapForNodeModules = async (
     tags,
     packageDescriptor,
   );
-  return translateGraph(packageLocation, moduleSpecifier, graph);
+  return translateGraph(packageLocation, moduleSpecifier, graph, tags);
 };

--- a/packages/compartment-mapper/src/parse.js
+++ b/packages/compartment-mapper/src/parse.js
@@ -1,7 +1,19 @@
+// @ts-check
+/// <reference types="ses" />
+
 import { parseExtension } from './extension.js';
 import * as json from './json.js';
 
 const { entries, freeze, fromEntries } = Object;
+
+/**
+ * @template {string | number} Key
+ * @template Value
+ * @param {Record<Key, Value>} object
+ * @param {Key} key
+ * @returns {boolean}
+ */
+const has = (object, key) => Object.prototype.hasOwnProperty.call(object, key);
 
 // q, as in quote, for enquoting strings in error messages.
 const q = JSON.stringify;
@@ -9,6 +21,7 @@ const q = JSON.stringify;
 // TODO: parsers should accept bytes and perhaps even content-type for
 // verification.
 
+/** @type {ParseFn} */
 export const parseMjs = (source, _specifier, location, _packageLocation) => {
   return {
     parser: 'mjs',
@@ -16,8 +29,13 @@ export const parseMjs = (source, _specifier, location, _packageLocation) => {
   };
 };
 
+/** @type {ParseFn} */
 export const parseJson = (source, _specifier, location, _packageLocation) => {
+  /** @type {Readonly<Array<string>>} */
   const imports = freeze([]);
+  /**
+   * @param {Object} exports
+   */
   const execute = exports => {
     exports.default = json.parse(source, location);
   };
@@ -27,15 +45,22 @@ export const parseJson = (source, _specifier, location, _packageLocation) => {
   };
 };
 
+/**
+ * @param {Object<string, ParseFn>} extensions - maps a file extension to the
+ * corresponding parse function.
+ * @param {Object<string, string>} types - In a rare case, the type of a module
+ * is implied by package.json and should not be inferred from its extension.
+ * @returns {ParseFn}
+ */
 export const makeExtensionParser = (extensions, types) => {
   return (source, specifier, location, packageLocation) => {
     let extension;
-    if (Object(types) === types && hasOwnProperty.call(types, specifier)) {
+    if (Object(types) === types && has(types, specifier)) {
       extension = types[specifier];
     } else {
       extension = parseExtension(location);
     }
-    if (!hasOwnProperty.call(extensions, extension)) {
+    if (!has(extensions, extension)) {
       throw new Error(
         `Cannot parse module ${specifier} at ${location}, no parser configured for that extension`,
       );
@@ -45,16 +70,23 @@ export const makeExtensionParser = (extensions, types) => {
   };
 };
 
+/** @type {Record<string, ParseFn>} */
 export const parserForLanguage = {
   mjs: parseMjs,
   json: parseJson,
 };
 
+/**
+ * @param {Object<string, ParserDescriptor>} parsers
+ * @param {Object<string, string>} types - In a rare case, the type of a module
+ * is implied by package.json and should not be inferred from its extension.
+ * @returns {ParseFn}
+ */
 export const mapParsers = (parsers, types) => {
   const parserForExtension = [];
   const errors = [];
   for (const [extension, language] of entries(parsers)) {
-    if (hasOwnProperty.call(parserForLanguage, language)) {
+    if (has(parserForLanguage, language)) {
       const parser = parserForLanguage[language];
       parserForExtension.push([extension, parser]);
     } else {

--- a/packages/compartment-mapper/src/search.js
+++ b/packages/compartment-mapper/src/search.js
@@ -1,3 +1,5 @@
+// @ts-check
+
 import { relativize } from './node-module-specifier.js';
 import { relative } from './url.js';
 
@@ -6,13 +8,29 @@ const q = JSON.stringify;
 
 const decoder = new TextDecoder();
 
+/**
+ * @param {string} rel - a relative URL
+ * @param {string} abs - a fully qualified URL
+ * @returns {string}
+ */
 const resolveLocation = (rel, abs) => new URL(rel, abs).toString();
 
-// Searches for the first ancestor directory of a module file that contains a
-// package.json.
-// Probes by attempting to read the file, not stat.
-// To avoid duplicate work later, returns the text of the package.json for
-// inevitable later use.
+/**
+ * Searches for the first ancestor directory of a module file that contains a
+ * package.json.
+ * Probes by attempting to read the file, not stat.
+ * To avoid duplicate work later, returns the text of the package.json for
+ * inevitable later use.
+ *
+ * @param {ReadFn} read
+ * @param {string} moduleLocation
+ * @returns {Promise<{
+ *   packageLocation: string,
+ *   packageDescriptorLocation: string,
+ *   packageDescriptorText: string,
+ *   moduleSpecifier: string,
+ * }>}
+ */
 export const search = async (read, moduleLocation) => {
   let directory = resolveLocation('./', moduleLocation);
   for (;;) {

--- a/packages/compartment-mapper/src/types.js
+++ b/packages/compartment-mapper/src/types.js
@@ -1,0 +1,169 @@
+// @ts-check
+
+// /////////////////////////////////////////////////////////////////////////////
+
+// The schema of a compartment map JSON file:
+
+/**
+ * A compartment map describes how to construct an application as a graph of
+ * Compartments, each corresponding to Node.js style packaged modules.
+ *
+ * @typedef {Object} CompartmentMapDescriptor
+ * @property {Array<string>} tags
+ * @property {EntryDescriptor} entry
+ * @property {Record<string, CompartmentDescriptor>} compartments
+ */
+
+/**
+ * The entry descriptor of a compartment map denotes the root module of an
+ * application and the compartment that contains it.
+ *
+ * @typedef {Object} EntryDescriptor
+ * @property {string} compartment
+ * @property {string} module
+ */
+
+/**
+ * A compartment descriptor corresponds to a single Compartment
+ * of an assembled Application and describes how to construct
+ * one for a given library or application package.json.
+ *
+ * @typedef {Object} CompartmentDescriptor
+ * @property {string} label
+ * @property {string} location
+ * @property {Record<string, ModuleDescriptor>} modules
+ * @property {Record<string, ScopeDescriptor>} scopes
+ * @property {Record<string, ParserDescriptor>} parsers - type for extension
+ * @property {Record<string, ParserDescriptor>} types - type for module specifier
+ */
+
+/**
+ * For every module explicitly mentioned in an `exports` field of a
+ * package.json, there is a corresponding module descriptor.
+ *
+ * @typedef {Object} ModuleDescriptor
+ * @property {string} [compartment]
+ * @property {string} [module]
+ * @property {string} [location]
+ * @property {ParserDescriptor} [parser]
+ * @property {string} [exit]
+ */
+
+/**
+ * Scope descriptors link all names under a prefix to modules in another
+ * compartment, like a wildcard.
+ * These are employed to link any module not explicitly mentioned
+ * in a package.json file, when that package.json file does not have
+ * an explicit `exports` map.
+ *
+ * @typedef {Object} ScopeDescriptor
+ * @property {string} compartment
+ * @property {string} [module]
+ */
+
+/**
+ * @typedef {'mjs' | 'cjs' | 'json'} ParserDescriptor
+ */
+
+// /////////////////////////////////////////////////////////////////////////////
+
+// IO capabilities and archives:
+
+/**
+ * @typedef {Object} ArchiveWriter
+ * @property {WriteFn} write
+ * @property {SnapshotFn} snapshot
+ */
+
+/**
+ * @callback WriteFn
+ * @param {string} name
+ * @param {Uint8Array} bytes
+ * @returns {Promise<void>}
+ */
+
+/**
+ * @typedef {Object} ArchiveReader
+ * @property {ReadFn} read
+ */
+
+/**
+ * @callback ReadFn
+ * @param {string} name
+ * @returns {Promise<Uint8Array>} bytes
+ */
+
+/**
+ * @typedef {Object} Application
+ * @property {ExecuteFn} import
+ */
+
+/**
+ * @callback ExecuteFn
+ * @param {ExecuteOptions} options
+ * @returns {Promise<Object>}
+ */
+
+/**
+ * @callback SnapshotFn
+ * @returns {Promise<Uint8Array>}
+ */
+
+// /////////////////////////////////////////////////////////////////////////////
+
+// Shared machinery for assembling applications:
+
+/**
+ * @callback ImportHookMaker
+ * @param {string} packageLocation
+ * @param {ParseFn} parse
+ * @returns {ImportHook}
+ */
+
+/**
+ * @callback ParseFn
+ * @param {string} source
+ * @param {string} specifier
+ * @param {string} location
+ * @param {string} packageLocation
+ * @returns {{
+ *   parser: ParserDescriptor,
+ *   record: FinalStaticModuleType,
+ * }}
+ */
+
+/**
+ * @typedef {Object} ExecuteOptions
+ * @property {Object} [globals]
+ * @property {Object} [globalLexicals]
+ * @property {Array<Transform>} [transforms]
+ * @property {Array<Transform>} [__shimTransforms__]
+ * @property {Record<string, string>} [modules]
+ * @property {typeof Compartment.prototype.constructor} [Compartment]
+ */
+
+/**
+ * @typedef {ExecuteOptions & Object} AssemblyOptions
+ * @property {AssembleImportHook} makeImportHook
+ */
+
+// /////////////////////////////////////////////////////////////////////////////
+
+// Communicating source files from an archive snapshot, from archive.js to
+// import-hook.js:
+
+/**
+ * @typedef {Record<string, CompartmentSources>} Sources
+ */
+
+/**
+ * @typedef {Record<string, ModuleSource>} CompartmentSources
+ */
+
+/**
+ * @typedef {Object} ModuleSource
+ * @property {string} [location]
+ * @property {Uint8Array} [bytes]
+ * @property {ParserDescriptor} [parser]
+ * @property {string} [exit]
+ */

--- a/packages/compartment-mapper/src/url.js
+++ b/packages/compartment-mapper/src/url.js
@@ -1,3 +1,5 @@
+// @ts-check
+
 // Derrived from https://github.com/junosuarez/url-relative
 //
 // Copyright (c) MMXV jden jason@denizac.org
@@ -16,6 +18,13 @@
 // OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 // CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
+/**
+ * Computes the relative URL from the referrer to the location.
+ *
+ * @param {string} referrer
+ * @param {string} location
+ * @returns {string}
+ */
 export const relative = (referrer, location) => {
   referrer = String(referrer || '');
   location = String(location || '');
@@ -27,15 +36,11 @@ export const relative = (referrer, location) => {
   const referrerURL = new URL(referrer);
   const locationURL = new URL(location);
 
-  if (referrerURL.host !== locationURL.host) {
-    return location;
-  }
-
-  if (referrerURL.protocol !== locationURL.protocol) {
-    return location;
-  }
-
-  if (referrerURL.auth !== locationURL.auth) {
+  if (
+    referrerURL.host !== locationURL.host ||
+    referrerURL.port !== locationURL.port ||
+    referrerURL.protocol !== locationURL.protocol
+  ) {
     return location;
   }
 

--- a/packages/compartment-mapper/src/zip.js
+++ b/packages/compartment-mapper/src/zip.js
@@ -1,17 +1,34 @@
-// Decouples Zip usage from JSZip's particular presentation.
+// @ts-check
+
+/**
+ * Decouples Zip usage from JSZip's particular presentation.
+ */
 
 import { ZipReader } from './zip/reader.js';
 import { ZipWriter } from './zip/writer.js';
 
+/**
+ * @param {Uint8Array} data
+ * @param {string} location
+ * @returns {Promise<ArchiveReader>}
+ */
 export const readZip = async (data, location) => {
   const reader = new ZipReader(data, { name: location });
+  /** @type {ReadFn} */
   const read = async path => reader.read(path);
   return { read };
 };
 
+/**
+ * @returns {ArchiveWriter}
+ */
 export const writeZip = () => {
   const writer = new ZipWriter();
-  const write = async (path, data) => writer.write(path, data);
+  /** @type {WriteFn} */
+  const write = async (path, data) => {
+    writer.write(path, data);
+  };
+  /** @type {SnapshotFn} */
   const snapshot = async () => writer.snapshot();
   return { write, snapshot };
 };

--- a/packages/compartment-mapper/src/zip/compression.js
+++ b/packages/compartment-mapper/src/zip/compression.js
@@ -1,2 +1,4 @@
+// @ts-check
+
 // STORE is the magic number for "not compressed".
 export const STORE = '\x00\x00';

--- a/packages/compartment-mapper/src/zip/format-writer.js
+++ b/packages/compartment-mapper/src/zip/format-writer.js
@@ -31,7 +31,6 @@
  * }} BufferWriter
  */
 
-import './types.js';
 import { crc32 } from './crc32.js';
 import * as signature from './signature.js';
 import * as compression from './compression.js';

--- a/packages/compartment-mapper/src/zip/signature.js
+++ b/packages/compartment-mapper/src/zip/signature.js
@@ -1,3 +1,4 @@
+// @ts-check
 /* eslint no-bitwise: ["off"] */
 
 /**

--- a/packages/compartment-mapper/src/zip/writer.js
+++ b/packages/compartment-mapper/src/zip/writer.js
@@ -1,5 +1,5 @@
 // @ts-check
-import './types.js';
+
 import { BufferWriter } from './buffer-writer.js';
 import { writeZip } from './format-writer.js';
 

--- a/packages/ses/NEWS.md
+++ b/packages/ses/NEWS.md
@@ -2,7 +2,9 @@ User-visible changes in SES:
 
 ## Next release
 
-- No changes yet
+- Expand TypeScript definitions to include Compartment, StaticModuleRecord,
+  StaticModuleType, RedirectStaticModuleInterface, FinalStaticModuleType,
+  ThirdPartyStaticModuleInterface, Transform, ImportHook, and ModuleMapHook.
 
 ## Release 0.12.3 (1-Mar-2021)
 

--- a/packages/ses/index.d.ts
+++ b/packages/ses/index.d.ts
@@ -5,15 +5,48 @@
 import type { Hardener } from '@agoric/make-hardener';
 import type { CompartmentConstructor } from './src/compartment-shim';
 import type { Lockdown } from './src/lockdown-shim';
+import type { StaticModuleRecord } from './module-shim';
 
 // For scripts.
 declare var harden: Hardener;
 declare var lockdown: Lockdown;
-declare var Compartment: CompartmentConstructor;
+declare var Compartment : ReturnType<CompartmentConstructor>;
+declare var StaticModuleRecord : StaticModuleRecord;
+
+declare type StaticModuleType = RedirectStaticModuleInterface | FinalStaticModuleType;
+declare interface RedirectStaticModuleInterface {
+  readonly record: FinalStaticModuleType,
+  specifier: string
+};
+declare type FinalStaticModuleType = StaticModuleRecord | ThirdPartyModuleInterface;
+declare interface ThirdPartyStaticModuleInterface {
+  readonly imports: Array<string>,
+  readonly execute: (exports: Object) => void,
+};
+
+declare type Transform = (source: string) => string;
+declare type ImportHook = (moduleSpecifier: string) => Promise<Object>;
+declare type ModuleMapHook = (moduleSpecifier: string) => string | Object | void;
 
 declare global {
   // For modules.
   var harden: Hardener;
   var lockdown : Lockdown;
-  var Compartment : CompartmentConstructor;
+  var Compartment : ReturnType<CompartmentConstructor>;
+  var StaticModuleRecord : StaticModuleRecord;
+
+  type StaticModuleType = RedirectStaticModuleInterface | FinalStaticModuleType;
+  interface RedirectStaticModuleInterface {
+    readonly record: FinalStaticModuleType,
+    specifier: string
+  };
+  type FinalStaticModuleType = StaticModuleRecord | ThirdPartyModuleInterface;
+  interface ThirdPartyStaticModuleInterface {
+    readonly imports: Array<string>,
+    readonly execute: (exports: Object) => void,
+  };
+
+  type Transform = (source: string) => string;
+  type ImportHook = (moduleSpecifier: string) => Promise<Object>;
+  type ModuleMapHook = (moduleSpecifier: string) => string | Object | void;
 }

